netket/utils/jax.py

```diff
@@ -84,3 +84,36 @@ def maybe_scalar_fun(apply_fun, pars, x, *args, **kwargs):
         return res
 
     return HashablePartial(maybe_scalar_fun, fun)
+
+
+def wrap_to_support_multiple_batch_axes(fun):
+    """
+    Wraps the flax-compatible apply function, assuming that the state input is the
+    second argumen with possibly many leading batch axes,
+    so that it always calls the wrapped function with a tensor with exactly 2 dimensions.
+    The result is reshaped so that it reflects the dimensions of the input.
+
+    DEVNOTE: This function is used because some users like to define their models with a vmap,
+    only supporting a single batch axis.
```

Comment:
I don't know how I feel about adding to netket 100 more lines, and a wrapper, to support a terrible habit.

We should rather instruct those 'users' in the right ways of `vectorize` their models rather than fix their bad behaviour ourself.

(I feel netket's getting too many magics...)

Reply:
netket has always corrected for this with the numba operators. It's just bizar not to do it now for the jax operator. This 'bad' behavior has already been encouraged by all previous versions of netket.
 It just introduced easy mistakes (it took me time to understand that this is what was happening). Furthermore, all other libaries do the same: you have a single batch axis...

Reply:
I would be in favor of having requirements on models / user inputs rather than putting too many conversions inside of netket if possible. For example, why can't we require users to have their models take a single batch axis only?

Reply:
@jwnys a while ago we had decided to explicitly not require that models support scalar inputs, but we never made a choice on whether we would require more than 1 batch dimension or not, even though I guess my wording does indeed suggest we only require a single batch axis support, so that's the defecto standard now.

What is missing are consistent tests to catch this.